### PR TITLE
Coverage instrumentation benchmark

### PR DESCRIPTION
**Pull request by fmeum**

### CODE REVIEW

**Comment by fmeum** [comment]
The results are in:
```
Benchmark                                                                  Mode  Cnt        Score       Error  Units
CoverageInstrumentationBenchmark.local_DirectByteBuffer_NeverZero         thrpt   25   530968.068 ± 61123.724  ops/s
CoverageInstrumentationBenchmark.staticMethod_DirectByteBuffer_NeverZero  thrpt   25   904792.671 ± 15894.743  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_NeverZero            thrpt   25  1341417.386 ± 13940.072  ops/s
CoverageInstrumentationBenchmark.uninstrumented                           thrpt   25  3191350.228 ± 93318.855  ops/s
```

**Comment by bertschneider** [comment]
Didn't not review the code yet, these are just the numbers:
```
Benchmark                                                                  Mode  Cnt        Score        Error  Units
CoverageInstrumentationBenchmark.local_DirectByteBuffer_NeverZero         thrpt   25   895186.166 ±  13294.400  ops/s
CoverageInstrumentationBenchmark.staticMethod_DirectByteBuffer_NeverZero  thrpt   25  1141211.132 ±  36699.694  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_NeverZero            thrpt   25  1864667.328 ±  58712.132  ops/s
CoverageInstrumentationBenchmark.uninstrumented                           thrpt   25  4583192.689 ± 215539.678  ops/s
```

**Comment by fmeum** [comment]
I tested a new contender that is even a bit faster than `UnsafeCoverageMap` by replacing `reportCoverage` with

```java
  public static void recordCoverage(final int id) {
    final long address = countersAddress + id;
    final byte counter = UNSAFE.getByte(address);
    UNSAFE.putByte(address, (byte) (counter == -1 ? 1 : counter + 1));
  }
```

I get the following assembly output
```
[Verified Entry Point]
  # {method} {0x00007f431119ff88} 'recordCoverage' '(I)V' in 'com/code_intelligence/jazzer/instrumentor/Unsafe2CoverageMap'
  # parm0:    rsi       = int
  #           [sp+0x20]  (sp of caller)
  0x00007f4320f06380:   sub    $0x18,%rsp
  0x00007f4320f06387:   mov    %rbp,0x10(%rsp)
  0x00007f4320f0638c:   movslq %esi,%r10
  0x00007f4320f0638f:   movabs $0x7f42bc0b3130,%r11
  0x00007f4320f06399:   add    %r10,%r11
  0x00007f4320f0639c:   movsbl (%r11),%r8d
  0x00007f4320f063a0:   cmp    $0xffffffff,%r8d
  0x00007f4320f063a4:   je     0x00007f4320f063bf
  0x00007f4320f063a6:   inc    %r8d
  0x00007f4320f063a9:   mov    %r8b,(%r11)
  0x00007f4320f063ac:   add    $0x10,%rsp
  0x00007f4320f063b0:   pop    %rbp
  0x00007f4320f063b1:   cmp    0x340(%r15),%rsp             ;   {poll_return}
  0x00007f4320f063b8:   ja     0x00007f4320f063c7
  0x00007f4320f063be:   retq   
  0x00007f4320f063bf:   mov    $0x1,%r8d
  0x00007f4320f063c5:   jmp    0x00007f4320f063a9
  0x00007f4320f063c7:   movabs $0x7f4320f063b1,%r10         ;   {internal_word}
  0x00007f4320f063d1:   mov    %r10,0x358(%r15)
  0x00007f4320f063d8:   jmpq   0x00007f4319449100           ;   {runtime_call SafepointBlob}
  0x00007f4320f063dd:   hlt    
  0x00007f4320f063de:   hlt    
  0x00007f4320f063df:   hlt    
[Exception Handler]
  0x00007f4320f063e0:   jmpq   0x00007f43194e8100           ;   {no_reloc}
[Deopt Handler Code]
  0x00007f4320f063e5:   callq  0x00007f4320f063ea
  0x00007f4320f063ea:   subq   $0x5,(%rsp)
  0x00007f4320f063ef:   jmpq   0x00007f43194483a0           ;   {runtime_call DeoptimizationBlob}
  0x00007f4320f063f4:   hlt    
  0x00007f4320f063f5:   hlt    
  0x00007f4320f063f6:   hlt    
  0x00007f4320f063f7:   hlt  
```

The benchmark result:
```
Benchmark                                                        Mode  Cnt        Score       Error  Units
CoverageInstrumentationBenchmark.staticMethod_Unsafe2_NeverZero  thrpt   25  1499826.540 ±  7437.095  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_NeverZero   thrpt   25  1396303.996 ±  8676.955  ops/s
CoverageInstrumentationBenchmark.uninstrumented                  thrpt   25  3133380.792 ± 55194.103  ops/s
```

@sebastianpoeplau remarked that conditional moves could be faster, but I can't get the JIT to emit them.

**Comment by kyakdan** [comment]
> I tested a new contender that is even a bit faster than `UnsafeCoverageMap` by replacing `reportCoverage` with
> 
> ```java
>   public static void recordCoverage(final int id) {
>     final long address = countersAddress + id;
>     final byte counter = UNSAFE.getByte(address);
>     UNSAFE.putByte(address, (byte) (counter == -1 ? 1 : counter + 1));
>   }
> ```
> 
> I get the following assembly output
> 
> ```
> [Verified Entry Point]
>   # {method} {0x00007f431119ff88} 'recordCoverage' '(I)V' in 'com/code_intelligence/jazzer/instrumentor/Unsafe2CoverageMap'
>   # parm0:    rsi       = int
>   #           [sp+0x20]  (sp of caller)
>   0x00007f4320f06380:   sub    $0x18,%rsp
>   0x00007f4320f06387:   mov    %rbp,0x10(%rsp)
>   0x00007f4320f0638c:   movslq %esi,%r10
>   0x00007f4320f0638f:   movabs $0x7f42bc0b3130,%r11
>   0x00007f4320f06399:   add    %r10,%r11
>   0x00007f4320f0639c:   movsbl (%r11),%r8d
>   0x00007f4320f063a0:   cmp    $0xffffffff,%r8d
>   0x00007f4320f063a4:   je     0x00007f4320f063bf
>   0x00007f4320f063a6:   inc    %r8d
>   0x00007f4320f063a9:   mov    %r8b,(%r11)
>   0x00007f4320f063ac:   add    $0x10,%rsp
>   0x00007f4320f063b0:   pop    %rbp
>   0x00007f4320f063b1:   cmp    0x340(%r15),%rsp             ;   {poll_return}
>   0x00007f4320f063b8:   ja     0x00007f4320f063c7
>   0x00007f4320f063be:   retq   
>   0x00007f4320f063bf:   mov    $0x1,%r8d
>   0x00007f4320f063c5:   jmp    0x00007f4320f063a9
>   0x00007f4320f063c7:   movabs $0x7f4320f063b1,%r10         ;   {internal_word}
>   0x00007f4320f063d1:   mov    %r10,0x358(%r15)
>   0x00007f4320f063d8:   jmpq   0x00007f4319449100           ;   {runtime_call SafepointBlob}
>   0x00007f4320f063dd:   hlt    
>   0x00007f4320f063de:   hlt    
>   0x00007f4320f063df:   hlt    
> [Exception Handler]
>   0x00007f4320f063e0:   jmpq   0x00007f43194e8100           ;   {no_reloc}
> [Deopt Handler Code]
>   0x00007f4320f063e5:   callq  0x00007f4320f063ea
>   0x00007f4320f063ea:   subq   $0x5,(%rsp)
>   0x00007f4320f063ef:   jmpq   0x00007f43194483a0           ;   {runtime_call DeoptimizationBlob}
>   0x00007f4320f063f4:   hlt    
>   0x00007f4320f063f5:   hlt    
>   0x00007f4320f063f6:   hlt    
>   0x00007f4320f063f7:   hlt  
> ```
> 
> The benchmark result:
> 
> ```
> Benchmark                                                        Mode  Cnt        Score       Error  Units
> CoverageInstrumentationBenchmark.staticMethod_Unsafe2_NeverZero  thrpt   25  1499826.540 ±  7437.095  ops/s
> CoverageInstrumentationBenchmark.staticMethod_Unsafe_NeverZero   thrpt   25  1396303.996 ±  8676.955  ops/s
> CoverageInstrumentationBenchmark.uninstrumented                  thrpt   25  3133380.792 ± 55194.103  ops/s
> ```
> 
> @sebastianpoeplau remarked that conditional moves could be faster, but I can't get the JIT to emit them.

We could give this variant a try. It does not contain any jumps but results in more assembly instructions:
```java
public static void recordCoverage(final int id) {
    final long address = countersAddress + id;
    final byte counter = UNSAFE.getByte(address);
    UNSAFE.putByte(address, (byte) (counter % 0xff + 1));
}
```

File: agent/src/jmh/java/com/code_intelligence/jazzer/instrumentor/Unsafe2CoverageMap.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  static {
+    Unsafe unsafe;
+    try {
+      Field f = Unsafe.class.getDeclaredField("theUnsafe");
+      f.setAccessible(true);
+      unsafe = (Unsafe) f.get(null);
+    } catch (IllegalAccessException | NoSuchFieldException e) {
+      e.printStackTrace();
+      System.exit(1);
```
Comment (bertschneider) [inline agent/src/jmh/java/com/code_intelligence/jazzer/instrumentor/Unsafe2CoverageMap.java:31]: Should we fall back to the `ByteBuffer` version if, for whatever reason, `Unsafe` can not be obtained?

**Review by bertschneider — APPROVED** [review]
Really good and interesting benchmarks! 

Now, which one should we use? We could also ship multiple ones and make the strategy configurable.

**Comment by fmeum** [comment]
The latest benchmark results:
```
CoverageInstrumentationBenchmark.local_DirectByteBuffer_NeverZero           thrpt   25   606359.097 ±  51464.275  ops/s
CoverageInstrumentationBenchmark.staticMethod_DirectByteBuffer2_NeverZero   thrpt   25   848459.679 ± 127822.610  ops/s
CoverageInstrumentationBenchmark.staticMethod_DirectByteBuffer_NeverZero    thrpt   25   806509.876 ± 131400.478  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_AtomicSimpleIncrement  thrpt   25   258652.693 ±    994.936  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_NeverZero              thrpt   25  1393052.150 ±  11020.556  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_NeverZero2             thrpt   25  1477061.469 ±   7437.856  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_NeverZeroBranchfree    thrpt   25  1269241.908 ±  11487.180  ops/s
CoverageInstrumentationBenchmark.staticMethod_Unsafe_SimpleIncrement        thrpt   25  1716329.628 ±  33942.194  ops/s
CoverageInstrumentationBenchmark.uninstrumented                             thrpt   25  3194175.390 ±  65778.337  ops/s
```

**Comment by fmeum** [comment]
I will rerun the benchmarks on my Mac, but it seems pretty clear that `Unsafe2` wins - after all, it generates essentially optimal ASM. I opened #302, where we can discuss how to get reliable access to Unsafe and whether we need a fallback for the real coverage map implementation.

**Pull request merged**